### ISSUE TITLE
32-UniControl: A Unified Diffusion Model for Controllable Visual Generation In the Wild

### ISSUE THREAD

**Issue by mjmaher987** [sharif-ml-lab/visual_compositional_generation_survey#32]
## Task: Read the Paper and Do the following 4 steps
1. Highlight the paper systematically (**green** for important notes - **blue** for very important notes that worth mentioning in the final survey - **red** for the sentences that we don’t understand or need further attention in the future) [In Zotero].
2. Write the Wikipedia summary of the paper in the GitHub issue (_comment in this issue_).
3. Extract the information in the [excel file](https://docs.google.com/spreadsheets/d/15_sh6TqtUvZIJo1-RZxQtsG0Nvg0vPG_IFKCWiGMjdY/edit?usp=sharing).
4. Add the related sentences to the [overleaf file](https://www.overleaf.com/8167246262ccxtrzztjvpz). Note that when you add the sentence, don’t forget to write the reference in the format of ”[ #reference number that is mentioned in the Google Doc - Name of the Author (who added this to the overleaf) ]” at the end of the sentence.

**Comment by geekn0rd** [comment]
## **Introduction and Objective:**

Visual generative foundation models such as Stable Diffusion show promise in creating high quality images when prompted with arbitrary language. However, they often fall short in generating images with spatial, structural, or geometric controls. To address this challenge, models have been trained using both visual condition images and text prompts. This paper introduces UniControl, which can support up to eight different kind of visual condition input. Unlike previous models that solely focused on one type of visual condition, UniControl allows for the inclusion of various visual conditions without the need to train multiple models. The paper proposes two innovative modules that effectively reduce the parameter size of the model. Also UniControl has shown some ability to do zero-shot tasks and generalization.

<img width="806" alt="Screenshot 2023-12-02 at 5 21 41 in the afternoon" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/70889826/382e25d0-14ad-47d4-879a-19d01b059523">


## **Methodology: MOE-style adapter and task-aware HyperNet**

To enable the model to take a wide range of visual conditions from different tasks, as well as the language prompt, they reformulate the training conditions and target pairs for UniControl. Specifically, suppose we have a dataset consisting of $K$ tasks :
$` D := \{ D_1 \cup · · · \cup D_K \}`$,  and for each task training set $D_i$, denote the training pairs by $`([c_{text}, c_{task}], I_c, x)`$, with $c_{task}$ being the task instruction that indicates the task type, $c_{text}$ being the language prompt describing the target image, $I_c$ being the visual conditions, and $x$ being the target image. With the additional task instruction, UniControl can differentiate visual conditions from different tasks.
Also they change the U-net to support their additional inputs and the noise function becomes $`\epsilon_{\theta}(z_t, t, c_{task}, c_{text}, I_c)`$ and the rest of loss function is the same as Stable Diffusion. They utilize classifier-free guidance to randomly drop 30% text prompts to enhance the controllability of input visual conditions.
They introduce two novel designed modules, MOE-style adapter and task-aware HyperNet to learn meta-knowledge across tasks and to overcome the misalignment of low-level features from different tasks.

#### MOE-style adapter:

The original Mixture-of-Experts (MOE) is a neural network architecture that involves dividing the network into multiple sub-networks or "experts", each responsible for different aspects of the input data. 
The MOE-Style Adapter is a modification inspired by the design of Mixture-of-Experts (MOEs) but tailored for a specific context. It incorporates a set of convolution modules to function as adapters. These adapters are designed to capture features related to various low-level visual conditions in the input data.
$F_{\text{Adapter}}(I_k^c) := \sum_{i=1}^{K} \delta_{i, k} \cdot F^{(i)}_{\text{Cov1}} \circ F^{(i)}_{\text{Cov2}}(I_k^c)$
When faced with a zero-shot task, the model can retrieve and utilize the adapters that were previously trained on highly related pre-training tasks. The adapters can be specialized for different tasks, enhancing adaptability. It is mentioned that each convolution module in the adapter has approximately 70K parameters.


####  Task-aware HyperNet:
In general, a HyperNet  is a type of neural network that generates the weights or parameters for another neural network. Instead of having fixed weights like a traditional neural network, a HyperNet dynamically produces the parameters of another network.
The term "task-aware" indicates that the HyperNet is capable of adapting its behavior based on the given task. $c_{task}$ is the description of the task instruction. The task-aware HyperNet projects the $c_{task}$ into a task embedding using a CLIPText encoder. The injection of the task embedding into the trainable copy of ControlNet is inspired by style modulation in StyleGAN2. The task embedding is multiplied to each zero-convolution layer of ControlNet. The length of the embedding matches the number of input channels of the zero-conv layer, and each element in the embedding is multiplied to the convolution kernel per input channel. This operation effectively modulates the behavior of the convolution layers based on the task embedding.

##  **Contributions:**
- UniControl is a unified model capable of handling various visual conditions
- They collect a new dataset for multi-condition visual generation with more than 20 million imagetext-condition triplets over nine distinct tasks across five categories. (MultiGen-20M)
- They conduct extensive experiments to demonstrate that the unified model UniControl outperforms each single-task controlled image generation, thanks to learning the intrinsic relationships between different visual conditions. 
- UniControl shows the ability to adapt to unseen tasks in a zero-shot manner, highlighting its versatility and potential for widespread adoption in the wild.

##  **Experiments/Results:**

#### Benchmark Models and Task-Specific ControlNet Models:
- UniControl is compared against task-specific ControlNet models, where six tasks overlap with those presented in ControlNet.
- UniControl is implemented and compared against these task-specific ControlNet models for fair evaluation.


#### Visual Comparison with ControlNet:
- Visual comparison across different tasks (Canny, HED, Depth, Normal, Segmentation, Openpose, Bounding Box, and Outpainting) consistently shows UniControl outperforming the baseline ControlNet model.
- UniControl exhibits superior visual quality, alignment with conditions, and better handling of diverse tasks.

#### User Study:
- A user study is conducted comparing UniControl with official ControlNet checkpoints on six tasks. It consistently outperforms ControlNet in all tasks, as indicated by user preferences.

#### Quantitative Evaluation:
- UniControl outperforms ControlNet in terms of image perceptual distance on multiple tasks.
- FID analysis demonstrates that UniControl consistently surpasses baseline methods, including classic single-task-controlled methods, across a range of tasks.

#### Ablation Study:
- The full-version UniControl, with both components, significantly outperforms ablations, highlighting the importance of these components.

#### Zero-Shot Generalization:
- UniControl showcases surprising zero-shot capabilities in hybrid conditions combination and unseen tasks generalization.
- Hybrid task combinations, involving depth and human skeleton or segmentation map and human skeleton, demonstrate UniControl's adaptability to complex conditions.
- UniControl successfully generalizes to unseen tasks like gray image colorization, image deblurring, and image inpainting, leveraging shared attributes and implicit correlations among pre-training and new tasks.



##  **Conclusion and Future Work:**
- UniControl, showcases impressive performance, yet it grapples with inherent limitations associated with diffusion-based image generation models.
 Primarily, these limitations stem from the nature of the training data, drawn from a subset of the Laion-Aesthetics datasets, leading to potential biases.
 -  Despite employing keyword and image-based filtering methods to mitigate bias, the model may still generate outputs that are biased or of lower fidelity. 
 -  Additionally, UniControl faces challenges when aiming for high-quality human-like outputs, especially in scenarios where precision and fidelity are crucial. 
 - On a broader scale, the model's performance could be further enhanced by access to improved open-source datasets, preventing the generation of biased, toxic, sexualized, or otherwise harmful content.
